File: python/binaryTreePostOrder.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None) -> None:
        self.val = val
        self.left = left
        self.right = right


def postorderTraversal(root: TreeNode | None) -> list:
    array = []
    postorder(root, array)
    return array


def postorder(root: TreeNode | None, array: list):
    if root is None:
        return

    postorder(root.left, array)
    postorder(root.right, array)
    array.append(root.val)


def postorderTraversalIterative(root: TreeNode | None) -> list:
    if root is None:
        return []

    array = []
    stack = []
    stack.append(root)

    while len(stack) != 0:
        node = stack.pop()
        array.append(node.val)
        if node.left is not None:
            stack.append(node.left)
        if node.right is not None:
            stack.append(node.right)

    i, j = 0, len(array) - 1

    while i < j:
        array[i], array[j] = array[j], array[i]
        i, j = i + 1, j - 1

    return array


node2 = TreeNode(val=3)
node1 = TreeNode(val=2)
root = TreeNode(val=1, left=node1, right=node2)
print(postorderTraversal(root))
print(postorderTraversalIterative(root))

File: go/BinaryTreePostOrder/BinaryTreePostOrder.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func postorderTraversal(root *TreeNode) []int {
	array := []int{}
	postOrder(root, &array)
	return array
}

func postOrder(root *TreeNode, array *[]int) {
	if root == nil {
		return
	}

	postOrder(root.Left, array)
	postOrder(root.Right, array)
	*array = append(*array, root.Val)
}

func postorderTraversalIterative(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}

	array := []int{}
	stack := []*TreeNode{}

	stack = append(stack, root)

	for len(stack) != 0 {
		node := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		array = append(array, node.Val)
		if node.Left != nil {
			stack = append(stack, node.Left)
		}

		if node.Right != nil {
			stack = append(stack, node.Right)
		}
	}

	for i, j := 0, len(array)-1; i < j; i, j = i+1, j-1 {
		array[i], array[j] = array[j], array[i]
	}

	return array
}

func main() {
	node2 := &TreeNode{Val: 3}
	node1 := &TreeNode{Val: 2}
	root := &TreeNode{Val: 1, Left: node1, Right: node2}
	fmt.Println(postorderTraversal(root))
	fmt.Println(postorderTraversalIterative(root))
}
